// @cybozu/license-manager/dist/analyze.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyze = void 0;
const picocolors_1 = __importDefault(require("picocolors"));
const aggregate_1 = require("./functions/aggregate");
const depsUtils_1 = require("./functions/depsUtils");
const loadConfigScript_1 = require("./functions/loadConfigScript");
const detectPackageManager_1 = require("./functions/detectPackageManager");
const analyze = async ({ configFilePath, ...cliArgs }) => {
    console.log("🔎 Analyzing dependencies...");
    const config = await (0, loadConfigScript_1.loadConfigScript)(configFilePath);
    const args = mergeConfig(cliArgs, config);
    const unreadLicenseTextPattern = [/.*/];
    const dependencies = await (0, aggregate_1.aggregate)(args.packageManager, args.query, args.cwd, args.workspace, unreadLicenseTextPattern, config);
    const errors = [];
    dependencies.forEach((dep) => {
        if (!isAllowLicense(dep, args)) {
            errors.push({
                name: dep.name,
                license: (0, depsUtils_1.getLicense)(dep),
            });
        }
    });
    if (errors.length) {
        console.error(picocolors_1.default.red("🚨 Unconfirmed license found"));
        console.error(errors);
        process.exit(1);
    }
    console.log(picocolors_1.default.green("✅ All dependencies confirmed"));
};
exports.analyze = analyze;
const mergeConfig = (args, config) => {
    return {
        query: args.query || config.analyze?.query || ":root *",
        cwd: args.cwd || config.cwd || "",
        workspace: args.workspace || config.workspace || "",
        allowLicenses: [...args.allowLicenses, ...(config.analyze?.allowLicenses || [])],
        allowPackages: [...args.allowPackages, ...(config.analyze?.allowPackages || [])],
        packageManager: args.packageManager || config.packageManager || (0, detectPackageManager_1.detectPackageManager)(),
    };
};
const isAllowLicense = (dep, args) => {
    if (args.allowLicenses.some((pattern) => (0, depsUtils_1.isMatchLicense)(dep, pattern)))
        return true;
    const isMatchAllowPack = args.allowPackages.find((allowPack) => {
        return (0, depsUtils_1.isMatchPackage)(dep, allowPack);
    });
    if (isMatchAllowPack)
        return true;
    return false;
};
