// @cybozu/license-manager/dist/functions/getLicenseText.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getApacheNotice = exports.getLicenseText = void 0;
const fs_1 = __importDefault(require("fs"));
const glob_1 = require("glob");
const picocolors_1 = __importDefault(require("picocolors"));
const fetcher_1 = require("./fetcher");
const getLicenseText = async (dep, overrideLicenseText) => {
    const depPath = dep.path;
    if (!depPath) {
        throw new Error("invalid dep: " + dep);
    }
    if (overrideLicenseText) {
        const result = overrideLicenseText(dep);
        if (typeof result === "string")
            return { licenseText: result };
        if (result !== undefined && "licenseText" in result) {
            return result.licenseText === undefined ? null : { licenseText: result.licenseText };
        }
        if (result !== undefined && "licensePageUrl" in result) {
            const licensePageUrl = result.licensePageUrl;
            if (licensePageUrl === undefined)
                return null;
            try {
                const res = await (0, fetcher_1.fetcher)(licensePageUrl);
                if (res.ok) {
                    console.log(picocolors_1.default.green(`✅ ${dep.name}@${dep.version} license page ${licensePageUrl} was successfully loaded`));
                    const licenseText = await res.text();
                    return { licenseText };
                }
                console.error(picocolors_1.default.red(`${dep.name}@${dep.version} license page ${licensePageUrl} not found`));
                return null;
            }
            catch (e) {
                console.error(picocolors_1.default.red(`${dep.name}@${dep.version} license page ${licensePageUrl} not accessed`));
                return null;
            }
        }
    }
    // TODO: read the file specified by "SEE LICENSE IN"
    const licensePattern = `${depPath}/LICEN@(S|C)E?(-MIT)?(.*)`;
    const res = glob_1.glob.sync(licensePattern, { nocase: true });
    if (res.length === 1) {
        return {
            licenseTextPath: res[0],
            licenseText: readFile(res[0]),
        };
    }
    return null;
};
exports.getLicenseText = getLicenseText;
const getApacheNotice = (dep) => {
    const depPath = dep.path;
    if (!depPath) {
        throw new Error("invalid dep: " + dep);
    }
    const noticePattern = `${depPath}/NOTICE?(.*)`;
    const res = glob_1.glob.sync(noticePattern, { nocase: true });
    if (res.length) {
        return {
            apacheNoticePath: res[0],
            apacheNotice: readFile(res[0]),
        };
    }
    return null;
};
exports.getApacheNotice = getApacheNotice;
const readFile = (licensePath) => {
    const text = fs_1.default.readFileSync(licensePath).toString("utf-8");
    return text.replace(/\r\n/g, "\n");
};
